### PR TITLE
Handle Throw Expressions

### PR DESCRIPTION
**Pull request by CharlesTaylor7**
This PR provides support for [Throw Expressions](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/throw#the-throw-expression).

### CODE REVIEW

File: test/CSharp7Features.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using System;
+
+namespace CSharp7
```
Comment (warrenseine) [inline test/CSharp7Features.cs]: Same as other PRs, no need for a new namespace / file.

File: test/CSharp7Features.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+{
+  class ThrowExpressions
+  {
+    public static int ThrowExpressionExample() =>
+    3 == 3 ? 2 : throw new Exception();
```
Comment (warrenseine) [inline test/CSharp7Features.cs]: I'd like to keep all commits passing as I tend to prefer rebase over merge to keep history readable. You should introduce this test with the appropriate feature support.

I should document it in a contribution guide.

File: test/AllInOne.cs [hunk 3]
```diff
@@ -774,6 +774,8 @@ class CSharp6Features
 
         private void NoOp()
         {
+            var foo = (2
```
Comment (warrenseine) [inline test/AllInOne.cs]: This change doesn't look related to throw expressions.

File: test/AllInOne.cs [hunk 4]
```diff
@@ -799,7 +801,7 @@ async void Test()
             s = $@"\{p.Name}
                                    ""\";
             s = $"Color [ R={func(b: 3):#0.##}, G={G:#0.##}, B={B:#0.##}, A={A:#0.##} ]";
-            
+
```
Comment (warrenseine) [inline test/AllInOne.cs]: It's ugly, but it's part of the test. Trailing spaces are expected to be removed by Prettier. But I realize it should be done only once in the file, and with an explicit comment.

File: src/parser.js [hunk 5]
```diff
@@ -32,8 +32,8 @@ function parseCSharp(text) {
   parser.addErrorListener(errorListener);
 
   const compilationUnit = parser.compilation_unit();
-  debugger;
   const result = simplifyTree(compilationUnit);
+  debugger;
```
Comment (warrenseine) [inline src/parser.js]: I would prefer not to see this in the history.

File: src/csharp/CSharpParser.g4 [hunk 6]
```diff
@@ -104,10 +104,12 @@ non_assignment_expression
 	| query_expression
 	| conditional_expression
 	| throw_expression
+	// | unary_expression
```
Comment (warrenseine) [inline src/csharp/CSharpParser.g4]: I would prefer not to see this in the history. Can you make sure your commits only contain non-debugging changes?

File: src/csharp/CSharpParser.g4 [hunk 7]
```diff
@@ -178,7 +178,6 @@ unary_expression
 	| '--' unary_expression
 	| OPEN_PARENS type CLOSE_PARENS unary_expression
 	| AWAIT unary_expression
-	| THROW unary_expression
```
Comment (warrenseine) [inline src/csharp/CSharpParser.g4]: Why is it unnecessary? Aren't you adding it?

File: test/AllInOne.Formatted.cs [hunk 8]
```diff
@@ -948,7 +948,6 @@ class CSharp6Features
 
         private void NoOp()
         {
-            var foo = (2, MethodCallExpression());
```
Comment (warrenseine) [inline test/AllInOne.Formatted.cs]: As mentioned on another commit, just rebase your code and simply do not introduce this line.

**Review by warrenseine — CHANGES_REQUESTED** [review]
No problem on the code itself, just some nit-picking over the batch of commits you're pushing.

File: test/CSharp7Features.cs (hunk 2 above)
Comment (CharlesTaylor7) [reply test/CSharp7Features.cs]: What I'm used to doing for work is to let Github handle automatically squashing & rebasing every PR into a single commit. You can configure any repo to do so: https://help.github.com/en/articles/configuring-commit-squashing-for-pull-requests

That said, I can manually rebase my PRs before submitting them if that's what you prefer.

File: src/csharp/CSharpParser.g4 (hunk 7 above)
Comment (CharlesTaylor7) [reply src/csharp/CSharpParser.g4]: Because you can throw more than just unary expressions. In the test case I'm throwing the result of an overloaded `+`. I added throw expressions under the non assignment expression node instead.

**Comment by CharlesTaylor7** [comment]
I have a rebased version of this that'll push after https://github.com/warrenseine/prettier-plugin-csharp/pull/20 gets merged (it's branched off of that one right now).

**Review by warrenseine — APPROVED** [review]
